File: elixir/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """
  @spec recite(strings :: [String.t()]) :: String.t()
  def recite([]), do: ""

  def recite(strings) do
    proverbs =
      strings
      |> Enum.chunk_every(2, 1, :discard)
      |> Enum.map(fn [a, b] -> "For want of a #{a} the #{b} was lost." end)

    (proverbs ++ ["And all for the want of a #{hd(strings)}.\n"])
    |> Enum.join("\n")
  end
end

File: go/proverb.go
package proverb

import "fmt"

func Proverb(rhyme []string) []string {
	if len(rhyme) == 0 {
		return []string{}
	}

	arr := make([]string, len(rhyme))
	for i := 1; i < len(rhyme); i++ {
		arr[i-1] = fmt.Sprintf("For want of a %s the %s was lost.", rhyme[i-1], rhyme[i])
	}

	arr[len(arr)-1] = fmt.Sprintf("And all for the want of a %s.", rhyme[0])

	return arr
}
